core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -186,8 +186,23 @@ private[spark] class TaskSetManager(
 
   // Add all our tasks to the pending lists. We do this in reverse order
   // of task index so that tasks with low indices get launched first.
-  for (i <- (0 until numTasks).reverse) {
-    addPendingTask(i)
+  addPendingTasks()
+
+  private def addPendingTasks(): Unit = {
+    val (_, duration) = Utils.timeTakenMs {
+      val hostToIndices = new HashMap[String, ArrayBuffer[Int]]()
+      for (i <- (0 until numTasks).reverse) {
+        addPendingTask(i, Option(hostToIndices))
+      }
+      // Convert preferred locations to racks in one invocation and zip with the origin indices.
+      // We de-duping the hosts to reduce this invocation further.
```

Comment:
Reword to

Resolve the rack for each host.  This can be somehow slow, so de-dupe the list of hosts, and assign the rack
to all relevant task indices.

Reply:
updated